### PR TITLE
Refactor OriginProof to be an Object

### PR DESCRIPTION
**Pull request by yuval-netanel**

### CODE REVIEW

File: KlarnaOnDemand/KODOriginProof.h [hunk 1]
```diff
@@ -14,6 +14,13 @@
  *
  *  @return An origin proof that needs to be sent along with the purchase for it to succeed.
  */
-+ (NSString *)generateWithAmount:(int)amount currency:(NSString *)currency userToken:(NSString *)userToken;
+- (id)initWithAmount:(int)amount currency:(NSString *)currency userToken:(NSString *)userToken;
+
+/**
+ *  Generates an origin proof NString that matches the details of the purchase it will be used to verify.
```
Comment (tomerlevi444) [inline KlarnaOnDemand/KODOriginProof.h]: `Generates an origin proof that matches the details of the order.
The origin proof is used by Klarna in order to verify that the order was originated by this device.`

File: KlarnaOnDemand/KODOriginProof.h [hunk 2]
```diff
@@ -14,6 +14,13 @@
...
+ *
+ *  @return An origin proof NString that needs to be sent along with the purchase for it to succeed.
+ */
```
Comment (tomerlevi444) [inline KlarnaOnDemand/KODOriginProof.h]: `An origin proof that matches the details of this order and identifies this device.`

File: KlarnaOnDemand/KODOriginProof.m [hunk 3]
```diff
@@ -3,12 +3,37 @@
...
+@interface KODOriginProof ()
+
+
+@property(nonatomic, assign) int amount;
+@property(strong, nonatomic) NSString *currency;
+@property(strong, nonatomic) NSString *userToken;
+
+@end
+
 @implementation KODOriginProof
```
Comment (tomerlevi444) [inline KlarnaOnDemand/KODOriginProof.m]: remove newline

File: KlarnaOnDemand/KODOriginProof.m [hunk 4]
```diff
@@ -3,12 +3,37 @@
 #import <Foundation/NSJSONSerialization.h>
 #import "NSData+Base64.h"
 
+@interface KODOriginProof ()
+
```
Comment (tomerlevi444) [inline KlarnaOnDemand/KODOriginProof.m]: redundant newline

**Comment by tomerlevi444** [comment]
Approved

File: KlarnaOnDemand/KODOriginProof.m (hunk 4 above)
Comment (yuval-netanel) [inline KlarnaOnDemand/KODOriginProof.m]: Done

File: KlarnaOnDemand/KODOriginProof.h (hunk 1 above)
Comment (yuval-netanel) [inline KlarnaOnDemand/KODOriginProof.h]: @grozen ?

File: KlarnaOnDemand/KODOriginProof.h (hunk 2 above)
Comment (yuval-netanel) [inline KlarnaOnDemand/KODOriginProof.h]: @grozen ?

File: KlarnaOnDemand/KODOriginProof.h (hunk 1 above)
Comment (grozen) [inline KlarnaOnDemand/KODOriginProof.h]: I have to dispute this description.
From a user's standpoint, this doesn't generate anything, this merely returns a string representation of something that was already generated (using `initWithAmount`).

File: KlarnaOnDemand/KODOriginProof.h (hunk 2 above)
Comment (grozen) [inline KlarnaOnDemand/KODOriginProof.h]: I actually prefer @yuval-netanel's description which is more to the point. I'd tweak it slightly to:

`@return A textual representation of the origin proof that needs to be sent along with the purchase for it to succeed.`

File: KlarnaOnDemand/KODOriginProof.m [hunk 5]
```diff
@@ -3,12 +3,37 @@
...
+  return self;
+}
+
+- (id)init {
+  NSAssert(NO, @"Initialize with -initWithAmount:currency:userToken");
+  return nil;
+}
+
+- (NSString *)description {
+  NSData *data = [self jsonDataWithDictionary:@{@"amount": [NSNumber numberWithInt: self.amount],
```
Comment (grozen) [inline KlarnaOnDemand/KODOriginProof.m]: I'd move the actual origin proof generation to the constructor. Store the result and return it when calling `description`.

Also, maybe align these key/value pairs somehow?

**Comment by grozen** [comment]
Approved

**Pull request merged**